src/watchdog.rs

```diff
@@ -0,0 +1,210 @@
+            }
+        }
+
+        #[cfg(esp_idf_version_major = "4")]
+        pub const fn new(duration: core::time::Duration, panic_on_trigger: bool) -> Self {
+            Self {
+                duration,
+                panic_on_trigger,
+            }
+        }
+    }
+
+    impl Default for WatchdogConfig {
+        fn default() -> Self {
+            Self {
+                duration: core::time::Duration::from_secs(CONFIG_ESP_TASK_WDT_TIMEOUT_S as u64),
+                panic_on_trigger: cfg!(esp_idf_esp_task_wdt_panic), // CONFIG_ESP_TASK_WDT_PANIC
+                #[cfg(not(esp_idf_version_major = "4"))]
+                subscribed_idle_tasks: {
+                    let subscribed_idle_tasks = heapless::Vec::new();
+                    if cfg!(esp_idf_esp_task_wdt_check_idle_task_cpu0) {
+                        subscribed_idle_tasks
+                            .push(crate::task::IdleTask::Core0)
+                            .unwrap();
+                    }
+                    #[cfg(any(esp32, esp32s3))]
+                    if cfg!(esp_idf_esp_task_wdt_check_idle_task_cpu1) {
+                        subscribed_idle_tasks
+                            .push(crate::task::IdleTask::Core1)
+                            .unwrap();
+                    }
+                    subscribed_idle_tasks
+                },
+            }
+        }
+    }
+
+    #[cfg(not(esp_idf_version_major = "4"))]
+    impl From<WatchdogConfig> for esp_task_wdt_config_t {
+        fn from(value: WatchdogConfig) -> Self {
+            esp_task_wdt_config_t {
+                timeout_ms: value.duration.as_millis() as u32,
+                trigger_panic: value.panic_on_trigger,
+                idle_core_mask: value
+                    .subscribed_idle_tasks
+                    .iter()
+                    .fold(0u32, |mask, core| mask & (1 << (*core as u32))),
+            }
+        }
+    }
+
+    pub trait Twdt {}
+
+    pub struct TWDTDriver<'d>(PhantomData<&'d mut ()>);
+
+    unsafe impl Send for TWDTDriver<'_> {}
+
+    static TWDT_DRIVER_REF_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+    impl Clone for TWDTDriver<'_> {
+        fn clone(&self) -> Self {
+            TWDT_DRIVER_REF_COUNT.fetch_add(1, core::sync::atomic::Ordering::Relaxed);
+            Self(Default::default())
+        }
+    }
+
+    impl<'d> TWDTDriver<'d> {
+        pub fn new(
+            _twdt: impl Peripheral<P = TWDT> + 'd,
+            config: &WatchdogConfig,
+        ) -> Result<Self, EspError> {
+            init_or_reconfigure(config)?;
+            Ok(Self(Default::default()))
+        }
+
+        // will fail in isr
+        pub fn watch_current_task(&mut self) -> Result<WatchdogSubscription<'_>, EspError> {
+            add_task(core::ptr::null_mut())?;
+            Ok(WatchdogSubscription::new())
+        }
+    }
+
+    impl Drop for TWDTDriver<'_> {
+        fn drop(&mut self) {
+            let _drop = TWDT_DRIVER_REF_COUNT
```

Comment:
> > The code is very-complex? Why not just fetch_sub?
> 
> `fetch_sub` will wrap at 0. If I check the counter at first if it's zero and then if I do `fetch_sub` in between, another thread could change the counter.

Not sure I understand:
```rust
let refcnt = TWDT_DRIVER_REF_COUNT.fetch_sub(1, Ordering::SeqCst);
if refcnt == 1 {
  // Deinit
} else if refcnt < 0 {
  unreachable!(); // Bug, should never happen
}
```

The whole point of atomics is that they serialize access and modifications to the variable.

> 
> > This is not de-initializing the driver when the count reaches 0
> 
> The problem is, the IDF subscribes the idle tasks on the watchdog. So deinit the driver here will fail because these tasks are subscribed.

Actually, an even bigger difficulty is that the watchdog might had been **started already** by ESP IDF itself using the `CONFIG_` kconfig values. Moreover, one or both the idle tasks might be subscribed already (as is the default).

I think we have to make a choice here:
* Option 1 - Whatever the initial state is, once `TWDTDriver` is created, user is in **full control of everything**, including  whether the watchdog runs on the idle tasks. For the watchdog on the idle tasks to continue to run, user **must** keep at least one `TWDTDriver` instance around. Moreover, dropping the last `TWDTDriver` ref should unsubscribe any idle tasks which are still subscribed and deinit the watchdog C TWDT driver
* Option 2 - `TWDTDriver` being active or not does not affect the policy towards idle task watchdogs. In other words, the user can create a `TWDTDriver`, reconfigure the idle tasks' watchdog behavior, then drop the `TWDTDriver` instance (and all clones), and the system will intelligently preserve the settings w.r.t. idle tasks and keep the watchdog C driver running, if necessary

I'm for now inclined to do "Option 1" as it is simpler to implement. Happy to hear your feedback of course!

Reply:
> Not sure I understand (corrected):

Thanks, didn't think of this solution.

> Actually, an even bigger difficulty is that the watchdog might had been started already by ESP IDF itself using the CONFIG_ kconfig values. Moreover, one or both the idle tasks might be subscribed already (as is the default).

Yes, thought on this as well.

Thoughts on the options:
- Option 1: 
  - To implement it this way, there must be the possibility of getting the idle task handles to unsubscribe them. I don't know the behavior of the idf if the user handles the un-subscription. I have to try. 
  - And secondly, is there a way that the idf does heavy work after the user app ends? If so, would be nice to watch this. 
  - Moreover, as you said, the user must drop the watchdog driver after everything else.
- Option 2: This was my first thought.

So, I would test if the watchdog is init at driver initialization and keep this state. At the end, the driver de-initializes the watchdog based on this state.

Reply:
I tested the behavior of different configs:

- CONFIG_TASK_WDT=y & CONFIG_TASK_WDT_CHECK_IDLE_TASK_CPU0=n/y & reinit by user: deinit watchdog from user no panic
- CONFIG_TASK_WDT=n & CONFIG_TASK_WDT_CHECK_IDLE_TASK_CPU0=n/y & init by user: deinit watchdog from user no panic

As a result, the watchdog can always be deinitialized. No need to check if initialized by idf. So, your option 1 is works fine.

Reply:
> esp32, esp32s2, esp32s3: xTaskGetIdleTaskHandleForCPU
esp32c3: xTaskGetIdleTaskHandle

Ah ok, after testing it I think the idle tasks are unsubscribed automatically by deinit the watchdog.

Reply:
> > esp32, esp32s2, esp32s3: xTaskGetIdleTaskHandleForCPU
> > esp32c3: xTaskGetIdleTaskHandle
> 
> Ah ok, after testing it I think the idle tasks are unsubscribed automatically by deinit the watchdog.
The reason why I'm pointing you at the APIs above is because with those it *is* possible to implement idle tasks' subscription even on older ESP IDFs.

> > What do you mean by "idf doing the heavy work"? What heavy work?
> 
> I mean if the IDF some how gets stuck after user task dropped everything and the watchdog needs to recover this.

I'm still not sure I understand what you mean. If you've seen any abnormal behavior, please let me know. Otherwise, FYI, ESP IDF does not and cannot "recover" anything, with or without watchdog running and watching (some) tasks. It can only ring the alarm for you (as in error messages in the console) and potentially panic, if you asked it to.

Reply:
> The reason why I'm pointing you at the APIs above is because with those it is possible to implement idle tasks' subscription/unsubcription even on older ESP IDFs.

Ah, you mean if the CONFIG_WATCH_IDLE_TASK is set to 0 and the user wants it to watch at runtime wee can do it for him? Good idea.